### PR TITLE
[ShapeInfer] FFTBase shape infer improvement - preserve input sizes bounds and labels

### PR DESCRIPTION
**Pull request by mitruska**
### Details:
 -  FFTBase - (DFT, IDFT) preserve input sizes bounds and labels,
 similar change for RDFT and IRDFT should be considered separately (details in the ticket).
 
- Additional suggestions from (https://github.com/openvinotoolkit/openvino/pull/19070)
  - remove `axes_are_known` variable
  - pass rank as `int64_t`

### Tickets:
 - 97219

### CODE REVIEW

File: src/core/shape_inference/include/fft_common_validation.hpp [hunk 1]
```diff
@@ -40,12 +40,12 @@ void validate_input_rank(const ov::op::util::FFTBase* op,
     if (fft_kind == FFTKind::RealInput) {
         NODE_SHAPE_INFER_CHECK(op,
                                input_shapes,
-                               input_rank >= static_cast<size_t>(axes_shape[0].get_length()),
+                               input_rank >= static_cast<int64_t>(axes_shape[0].get_length()),
```
Comment (praasz) [inline src/core/shape_inference/include/fft_common_validation.hpp:43]:
```suggestion
                               ov::cmp::ge(input_rank, axes_shape[0].get_length()),
```

Comment (mitruska) [reply src/core/shape_inference/include/fft_common_validation.hpp]: Ok, done:)

File: src/core/shape_inference/include/fft_common_validation.hpp [hunk 2]
```diff
@@ -124,16 +123,18 @@ void validate_signal_size(const ov::op::util::FFTBase* op,
 template <class T>
 void shape_validation(const ov::op::util::FFTBase* op,
                       const std::vector<T>& input_shapes,
-                      std::vector<int64_t>& axes,
-                      bool axes_are_known,
+                      std::unique_ptr<std::vector<int64_t>>& axes,
```
[the remark this answers is not in the document]
Comment (mitruska) [reply src/core/shape_inference/include/fft_common_validation.hpp]: Hm okay, changed to pass raw pointer

File: src/core/shape_inference/include/fft_base_shape_inference.hpp [hunk 3]
```diff
@@ -2,57 +2,90 @@
...
+    if (const auto output_bounds = get_input_bounds<TRShape, int64_t>(op, 2, ta)) {
+        constexpr auto minus_one_bound = std::make_pair(dim::inf_bound, dim::inf_bound);
+        auto labels =
+            op->get_input_size() > 2 ? op->get_input_source_output(2).get_tensor().get_value_label() : TensorLabel();
+        const bool propagate_labels = !labels.empty();
+        const auto num_of_axes = axes.size();
+        for (size_t i = 0; i < num_of_axes; ++i) {
+            if ((*output_bounds)[i] != minus_one_bound) {
+                output_shape[(axes)[i]] = DimType((*output_bounds)[i].first, (*output_bounds)[i].second);
+                if (propagate_labels) {
```
Comment (praasz) [inline src/core/shape_inference/include/fft_base_shape_inference.hpp:34]: What if axes are labeled like {1, no_label,2}? The set label will throw exception for `no_label`.

File: src/core/shape_inference/include/fft_common_validation.hpp [hunk 4]
```diff
@@ -73,8 +72,8 @@ void validate_axes(const ov::op::util::FFTBase* op,
     // according to the RDFT operation specification, axes should be integers from -r to (r - 1)
     // inclusively, where r = rank(data). A negative axis 'a' is interpreted as an axis 'r + a'.
     const int64_t axis_correction = (fft_kind == FFTKind::RealInput) ? input_rank : (input_rank - 1);
-    auto axis_min_value = -static_cast<int64_t>(input_rank);
-    auto axis_max_value = static_cast<int64_t>(input_rank) - 1;
+    auto axis_min_value = -input_rank;
+    auto axis_max_value = input_rank - 1;
```
Comment (praasz) [inline src/core/shape_inference/include/fft_common_validation.hpp:76]:
```cpp
    ov::normalize_axes(op, axis_correction, axes);
    auto axes_set = AxisSet(std::vector<size_t>(axes.begin(), axes.end()));
    NODE_VALIDATION_CHECK(op, axes.size() == axes_set.size(), "Each axis must be unique.");
```
It looks like exceptions are thrown correctly in test, only message message is different.

Can you check if loop below can be replaced?

Comment (mitruska) [reply src/core/shape_inference/include/fft_common_validation.hpp]: Okay, looks legit, replaced.

File: src/core/shape_inference/include/fft_base_shape_inference.hpp (hunk 3 above)
Comment (mitruska) [reply src/core/shape_inference/include/fft_base_shape_inference.hpp]: Good catch, actually not the labels of the `axes`, but the `sizes`, but thanks, fixed.

**Review by praasz — APPROVED** [review]

**Pull request merged**